src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
My thinking was that it might make sense to try and mimic [TransactNamedPipe](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365790%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396). That also says in case of overflow that you can get the rest of the data by reading.
Retrying will only work properly if the operation in question is idempotent. Any ioctl that has a side-effect can not necessarily just be attempted again.